Sources/Common/Core/DataArray/index.js

```diff
@@ -313,42 +329,51 @@ function vtkDataArray(publicAPI, model) {
 // Object factory
 // ----------------------------------------------------------------------------
 
-const DEFAULT_VALUES = {
-  name: '',
-  numberOfComponents: 1,
-  size: 0,
-  dataType: DefaultDataType,
-  rangeTuple: [0, 0],
-  // values: null,
-  // ranges: null,
-};
+function defaultValues(initialValues) {
+  return {
+    name: '',
+    numberOfComponents: 1,
+    size: 0,
+    dataType: DefaultDataType,
+    rangeTuple: [0, 0],
+    // data: macro.newTypedArray(DefaultValues),
+    // ranges: null,
+    ...initialValues,
+  };
+}
 
 // ----------------------------------------------------------------------------
 
 export function extend(publicAPI, model, initialValues = {}) {
-  Object.assign(model, DEFAULT_VALUES, initialValues);
-
-  if (!model.empty && !model.values && !model.size) {
+  if (!initialValues.empty && !initialValues.data && !initialValues.size) {
     throw new TypeError(
-      'Cannot create vtkDataArray object without: size > 0, values'
+      'Cannot create vtkDataArray object without: size > 0, data'
     );
   }
-
-  if (!model.values) {
-    model.values = macro.newTypedArray(model.dataType, model.size);
-  } else if (Array.isArray(model.values)) {
-    model.values = macro.newTypedArrayFrom(model.dataType, model.values);
-  }
-
-  if (model.values) {
-    model.size = model.values.length;
-    model.dataType = getDataType(model.values);
-  }
+  delete initialValues.empty;
+  Object.assign(initialValues, defaultValues(initialValues));
 
   // Object methods
   macro.obj(publicAPI, model);
   macro.set(publicAPI, model, ['name', 'numberOfComponents']);
 
+  model.dataType = initialValues.dataType
+    ? initialValues.dataType
+    : DefaultDataType;
+  if (!initialValues.data) {
+    initialValues.data = macro.newTypedArray(
+      model.dataType,
+      initialValues.size
+    );
+  } else if (Array.isArray(initialValues.data)) {
+    initialValues.data = macro.newTypedArrayFrom(
+      model.dataType,
+      initialValues.data
+    );
+  }
+  // Sort given values so that setData is called lastly in newInstance
+  sortProperties(initialValues);
+
```

Comment:
I think it's best to let the user put the `values` as parameter, since when the instance is created the array passed to constructor will be called that way. It seems strange to me that one pass a parameter to instanciation that is called after another way. We can support both too.

Reply:
I think it's best to let the user put the `values` as parameter, since when the instance is created the array passed to constructor will be called that way. It seems strange to me that one pass a parameter to instanciation that is called after another way. We can support both too.